File: telegram-bot-core/src/main/java/ru/tggc/telegrambotcore/annotation/handle/HandleMeta.java
package ru.tggc.telegrambotcore.annotation.handle;

import ru.tggc.telegrambotcore.dto.UserRole;

public record HandleMeta(UserRole[] requiredRoles, boolean canPublic, boolean canPrivate) {

    public static HandleMeta from(CallbackHandle a) {
        return new HandleMeta(a.requiredRoles(), a.canPublic(), a.canPrivate());
    }

    public static HandleMeta from(MessageHandle a) {
        return new HandleMeta(a.requiredRoles(), a.canPublic(), a.canPrivate());
    }

    public static HandleMeta from(PhotoHandle a) {
        return new HandleMeta(a.requiredRoles(), a.canPublic(), a.canPrivate());
    }

    public static HandleMeta from(CommandHandle a) {
        return new HandleMeta(a.requiredRoles(), a.canPublic(), a.canPrivate());
    }

    public static HandleMeta fromDefault() {
        return new HandleMeta(new UserRole[0], true, true);
    }
}

File: telegram-bot-core/src/main/kotlin/ru/tggc/telegrambotcore/annotation/handle/HandleMeta.kt
package ru.tggc.telegrambotcore.annotation.handle

import ru.tggc.telegrambotcore.dto.Access
import ru.tggc.telegrambotcore.dto.UserRole

data class HandleMeta(
    val requiredRoles: Array<UserRole> = emptyArray<UserRole>(),
    val canPublic: Boolean = true,
    val canPrivate: Boolean = true,
    val access: Access = Access.OWNER_ONLY,
) {
    companion object {
        fun from(a: CallbackHandle): HandleMeta {
            return HandleMeta(a.requiredRoles, a.canPublic, a.canPrivate, a.access)
        }

        fun from(a: MessageHandle): HandleMeta {
            return HandleMeta(a.requiredRoles, a.canPublic, a.canPrivate, a.access)
        }

        fun from(a: PhotoHandle): HandleMeta {
            return HandleMeta(a.requiredRoles, a.canPublic, a.canPrivate, a.access)
        }

        fun from(a: CommandHandle): HandleMeta {
            return HandleMeta(a.requiredRoles, a.canPublic, a.canPrivate, a.access)
        }

        fun from(a: TextHandle): HandleMeta {
            return HandleMeta(access = a.access)
        }

        fun from(): HandleMeta {
            return HandleMeta(access = Access.ANYONE)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as HandleMeta

        if (canPublic != other.canPublic) return false
        if (canPrivate != other.canPrivate) return false
        if (!requiredRoles.contentEquals(other.requiredRoles)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = canPublic.hashCode()
        result = 31 * result + canPrivate.hashCode()
        result = 31 * result + (requiredRoles.contentHashCode())
        return result
    }
}
